internal/ionoscloud/ionoscloud.go

```diff
@@ -0,0 +1,340 @@
+	allRecords, err := p.readAllRecords(ctx)
+	if err != nil {
+		return nil, err
+	}
+	epCollection := ionos.NewEndpointCollection[sdk.RecordRead](allRecords,
+		func(recordRead sdk.RecordRead) *endpoint.Endpoint {
+			record := *recordRead.GetProperties()
+			return endpoint.NewEndpointWithTTL(*record.GetName(), *record.GetType(), endpoint.TTL(*record.GetTtl()), *record.GetContent())
+		}, func(recordRead sdk.RecordRead) string {
+			record := *recordRead.GetProperties()
+			return *record.GetName() + "/" + *record.GetType() + "/" + strconv.Itoa(int(*record.GetTtl()))
+		})
+	return epCollection.RetrieveEndPoints(), nil
+}
+
+func (p *Provider) ApplyChanges(ctx context.Context, changes *plan.Changes) error {
+	epToCreate, epToDelete := ionos.GetCreateDeleteSetsFromChanges(changes)
+	zt, err := p.createZoneTree(ctx)
+	if err != nil {
+		return err
+	}
+	recordsToDelete := ionos.NewRecordCollection[sdk.RecordRead](epToDelete, func(ep *endpoint.Endpoint) []sdk.RecordRead {
+		logger := log.WithField(logFieldRecordName, ep.DNSName)
+		records := make([]sdk.RecordRead, 0)
+		zone := zt.FindZoneByDomainName(ep.DNSName)
+		if zone.Id == nil {
+			logger.Error("no zone found for record")
+			return records
+		}
+		logger = logger.WithField(logFieldZoneID, *zone.GetId())
+		zoneRecordReadList, err := p.client.GetRecordsByZoneIdAndName(ctx, *zone.GetId(), ep.DNSName)
+		if err != nil {
+			logger.Errorf("failed to get records for zone, error: %v", err)
+			return records
+		}
+		if !zoneRecordReadList.HasItems() {
+			logger.Warn("no records found to delete for zone")
+			return records
+		}
+		result := make([]sdk.RecordRead, 0)
+		for _, recordRead := range *zoneRecordReadList.GetItems() {
+			record := *recordRead.GetProperties()
+			if *record.GetType() == ep.RecordType {
+				for _, target := range ep.Targets {
+					if *record.GetContent() == target {
+						result = append(result, recordRead)
+					}
+				}
+			}
+		}
+		if len(result) == 0 {
+			logger.Warnf("no records in zone fit to delete for endpoint: %v", ep)
+		}
+		return result
+	})
+
+	if err := recordsToDelete.ForEach(func(recordRead sdk.RecordRead) error {
+		domainName := *recordRead.GetProperties().GetName()
+		zone := zt.FindZoneByDomainName(domainName)
+		if !zone.HasId() {
+			return fmt.Errorf("no zone found for domain '%s'", domainName)
+		}
+		err := p.client.DeleteRecord(ctx, *zone.GetId(), *recordRead.GetId())
+		return err
+	}); err != nil {
+		return err
+	}
+
+	recordsToCreate := ionos.NewRecordCollection[*sdk.RecordCreate](epToCreate, func(ep *endpoint.Endpoint) []*sdk.RecordCreate {
+		logger := log.WithField(logFieldRecordName, ep.DNSName)
+		zone := zt.FindZoneByDomainName(ep.DNSName)
+		if !zone.HasId() {
+			logger.Warnf("no zone found for domain '%s', skipping record creation", ep.DNSName)
+			return nil
+		}
+		result := make([]*sdk.RecordCreate, 0)
+		for _, target := range ep.Targets {
+			record := sdk.NewRecord(ep.DNSName, ep.RecordType, target)
```

Comment:
One thing I notice in both cloud and core provider is that handling of the priority of the priority of the SRV records which is stored in the endpoint target. 
`index := strings.Index(target, " ") `
`priority64, err := strconv.ParseInt(target[:index], 10, 64)`

Reply:
ok, however in order get that work, we must also  **put** the priority in the `endpoint.Target` see [line 224](https://github.com/ionos-cloud/external-dns-ionos-webhook/blob/fe200677231a093fdc1539b2911ecdca71dad4c8/internal/ionoscloud/ionoscloud.go#L224)
right?
Am I supposed to do this just for recordType `SRV` or regardless of the type?
When creating the record and I don't find a substring in the `endpoint.Target` which can be interpreted as a priority, i simply don't set the priority attribute on the record, right?

Reply:
Yes, we need to put the priority in the endpoint target, and this is only needed for the SRV type because the priority field is mandatory for this type of record. 
Example of how we do it in internal external dns:
`
target := *record.Content
if *record.Type == onedns.RecordTypeSRV {
    target = strconv.Itoa(*record.Prio) + " " + target
}
`

Currently the target is only interpreted as content of the record as I can see from the code.

Reply:
Not being able to add priority on record creation lead to additional complexity, but as far as I can see I think the case is handle well.